biodatasets/twadrl/twadrl.py

```diff
@@ -0,0 +1,167 @@
+mapped to how they are formally written in medical ontologies (SNOMED-CT and AMT). \
+"""
+
+_HOMEPAGE = "https://zenodo.org/record/55013"
+
+_LICENSE = "CC BY 4.0"
+
+_URLs = "https://zenodo.org/record/55013/files/datasets.zip"
+
+_SUPPORTED_TASKS = [Tasks.NAMED_ENTITY_RECOGNITION, Tasks.NAMED_ENTITY_DISAMBIGUATION]
+_SOURCE_VERSION = "1.0.0"
+_BIGBIO_VERSION = "1.0.0"
+
+
+class TwADRL(datasets.GeneratorBasedBuilder):
+    """TwADR-L: Dataset for Normalising Medical Concepts on Twitter."""
+
+    DEFAULT_CONFIG_NAME = "twadrl_source"
+    SOURCE_VERSION = datasets.Version(_SOURCE_VERSION)
+    BIGBIO_VERSION = datasets.Version(_BIGBIO_VERSION)
+
+    BUILDER_CONFIGS = [
+        BigBioConfig(
+            name="twadrl_source",
+            version=SOURCE_VERSION,
+            description="TwADR-L source schema",
+            schema="source",
+            subset_id="twadrl",
+        ),
+        BigBioConfig(
+            name="twadrl_bigbio_kb",
+            version=BIGBIO_VERSION,
+            description="TwADR-L simplified BigBio schema",
+            schema="bigbio_kb",
+            subset_id="twadrl",
+        ),
+    ]
+
+    def _info(self):
+        if self.config.schema == "source":
+            features = datasets.Features(
+                {
+                    "fold": datasets.Value("int32"),
+                    "cui": datasets.Value("string"),
+                    "medical_concept": datasets.Value("string"),
+                    "social_media_text": datasets.Value("string"),
+                }
+            )
+        elif self.config.schema == "bigbio_kb":
+            features = schemas.kb_features
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=features,
+            supervised_keys=None,
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        dl_dir = dl_manager.download_and_extract(_URLs)
+        dataset_dir = os.path.join(dl_dir, "datasets", "TwADR-L")
+        return [
+            datasets.SplitGenerator(
+                name=split,
+                gen_kwargs={
+                    "filepaths": glob.glob(
+                        os.path.join(dataset_dir, f"TwADR-L.fold-*.{split}.txt")
+                    ),
+                    "split": split,
+                },
+            )
+            for split in [
+                datasets.Split.TRAIN,
+                datasets.Split.VALIDATION,
+                datasets.Split.TEST,
+            ]
+        ]
+
+    def _generate_examples(self, filepaths, split):
+        for filepath in filepaths:
+            fold_id = re.search("TwADR-L\.fold-(\d)\.", filepath).group(1)
+            with open(filepath, "r", encoding="latin-1") as f:
+                document_id = f"{split}_{fold_id}"
+                for i, line in enumerate(f):
+                    id = f"{document_id}_{fold_id}_{i}"
```

Comment:
This is a good way to preserve the fold information. I am not sure what the consensus is on the k-fold datasets. I;ll ask around

Reply:
Based on Discord, have updated dataloader so that each of the k folds is its own split (eg "train_0", "train_1", ..., "train_k")